### ISSUE TITLE
Contextualize example doesn't seem to work

### ISSUE THREAD

**Issue by charleskawczynski** [vchuravy/GPUifyLoops.jl#28]
I just tried the contextualize example, and I'm getting this:

```
GPUifyLoops.jl>julia --project
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.0.0 (2018-08-08)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> include("examples/contextualize.jl")
[ Info: Recompiling stale cache file C:\Users\kawcz\.julia\compiled\v1.0\GPUifyLoops\BxvPf.ji for GPUifyLoops [ba82f77b-6841-5d2e-bd9f-4daf811aec27]
WARNING: Method definition overdub(Cassette.Context{N, M, T, P, B, H} where H<:Union{Cassette.DisableHooks, Nothing} where B<:Union{Nothing, Base.IdDict{Module, Base.Dict{Symbol, Cassette.BindingMeta}}} where P<:Cassette.AbstractPass where T<:Union{Nothing, Cassette.Tag{N, X, E} where E where X where N<:Cassette.AbstractContextName} where M where N<:Cassette.AbstractContextName, Any...) in module Cassette at C:\Users\kawcz\.julia\packages\Cassette\1rVkq\src\overdub.jl:500 overwritten in module GPUifyLoops at C:\Users\kawcz\.julia\packages\Cassette\1rVkq\src\overdub.jl:500.
WARNING: Method definition recurse(Cassette.Context{N, M, T, P, B, H} where H<:Union{Cassette.DisableHooks, Nothing} where B<:Union{Nothing, Base.IdDict{Module, Base.Dict{Symbol, Cassette.BindingMeta}}} where P<:Cassette.AbstractPass where T<:Union{Nothing, Cassette.Tag{N, X, E} where E where X where N<:Cassette.AbstractContextName} where M where N<:Cassette.AbstractContextName, Any...) in module Cassette at C:\Users\kawcz\.julia\packages\Cassette\1rVkq\src\overdub.jl:512 overwritten in module GPUifyLoops at C:\Users\kawcz\.julia\packages\Cassette\1rVkq\src\overdub.jl:512.
[ Info: Recompiling stale cache file C:\Users\kawcz\.julia\compiled\v1.0\CuArrays\7YFE0.ji for CuArrays [3a865a2d-5b23-5a0f-bc46-62713ec82fae]
┌ Warning: calls to Base intrinsics might be GPU incompatible
│   exception =
│    You called sin(x::T) where T<:Union{Float32, Float64} in Base.Math at special/trig.jl:30, maybe you intended to call sin(x::Float32) in CUDAnative at C:\Users\kawcz\.julia\packages\CUDAnative\Mdd3w\src\device\libdevice.jl:13 instead?
│    Stacktrace:
│     [1] sin at special/trig.jl:30
│     [2] #24 at C:\Users\kawcz\Dropbox\Caltech\work\dev\GPUifyLoops.jl\src\context.jl:127
│     [3] kernel! at C:\Users\kawcz\Dropbox\Caltech\work\dev\GPUifyLoops.jl\examples\contextualize.jl:8
└ @ CUDAnative C:\Users\kawcz\.julia\packages\CUDAnative\Mdd3w\src\compiler\irgen.jl:68
┌ Warning: calls to Base intrinsics might be GPU incompatible
│   exception =
│    You called sin(x::T) where T<:Union{Float32, Float64} in Base.Math at special/trig.jl:30, maybe you intended to call sin(x::Float32) in CUDAnative at C:\Users\kawcz\.julia\packages\CUDAnative\Mdd3w\src\device\libdevice.jl:13 instead?
│    Stacktrace:
│     [1] sin at special/trig.jl:30
│     [2] #24 at C:\Users\kawcz\Dropbox\Caltech\work\dev\GPUifyLoops.jl\src\context.jl:127
│     [3] kernel! at C:\Users\kawcz\Dropbox\Caltech\work\dev\GPUifyLoops.jl\examples\contextualize.jl:8
└ @ CUDAnative C:\Users\kawcz\.julia\packages\CUDAnative\Mdd3w\src\compiler\irgen.jl:68
```

Is there some other way that I need to launch Julia to get this working?

**Comment by vchuravy** [comment]
I strongly suspect that that is a false positive due to the pass putting a self-call in the reflected function body. https://github.com/vchuravy/GPUifyLoops.jl/blob/a59a88f5b2609463246a7f26dc3651ffc9baf359/src/context.jl#L46 which the verify still sees and inspect.
The self-call is dead code in the end, but the verifier at https://github.com/JuliaGPU/CUDAnative.jl/blob/e28c5f07cd2dae78a41d0d67ea2dfd4b374a92fe/src/compiler/irgen.jl#L59 doesn't see that.

**Comment by vchuravy** [comment]
You should be able to comment out https://github.com/vchuravy/GPUifyLoops.jl/blob/a59a88f5b2609463246a7f26dc3651ffc9baf359/src/context.jl#L39-L47 and then you should no longer get the verifier warning.

**Comment by charleskawczynski** [comment]
> You should be able to comment out
> 
> [GPUifyLoops.jl/src/context.jl](https://github.com/vchuravy/GPUifyLoops.jl/blob/a59a88f5b2609463246a7f26dc3651ffc9baf359/src/context.jl#L39-L47)
> 
> Lines 39 to 47 in [a59a88f](/vchuravy/GPUifyLoops.jl/commit/a59a88f5b2609463246a7f26dc3651ffc9baf359)
> 
>  self = GlobalRef(ref.method.module, ref.method.name) 
>  unknowably_false = GlobalRef(@__MODULE__, :unknowably_false) 
>  Cassette.insert_statements!(CI.code, CI.codelocs, 
>    (x, i) -> i == 1 ?  4 : nothing, 
>    (x, i) -> i == 1 ? [ 
>        Expr(:call, Expr(:nooverdub, unknowably_false)), 
>        Expr(:gotoifnot, Core.SSAValue(i), i+3), 
>        Expr(:call, Expr(:nooverdub, self), [Core.SlotNumber(i) for i in 2:length(CI.slotnames)]...), 
>        x] : [x]) 
> and then you should no longer get the verifier warning.

That worked, thanks!

**Comment by vchuravy** [comment]
To summarize a discussion I had with Charlie just now.

In order to make this work as expected:
```julia
f(x) = 2*x
g(x) = GPUifyLoops.contextualize(CUDA(), f)(x)
@assert g(3.0) == 6.0
f(x) = 3*x
@assert g(3.0) == 9.0
```

We have to place a fake call to the CPU version of `f`. That CPU version will contain incompatible code and the verifier correctly detects that :/ 
Will have to think in the context of https://github.com/JuliaGPU/CUDAnative.jl/pull/334 how to fix this properly.

In the long term my hope is that we can  avoid this whole dance, by inserting forward-edges but that needs work from the compiler side of things.

**Comment by vchuravy** [comment]
Fixed by #20
